### DISCUSSION TITLE
Constraints on the initial condition

### DISCUSSION THREAD

**Discussion by patil-apurva** [NVlabs/curobo#342] [Research Q&A]
In the cuRobo report it is mentioned that the first and last three states of the trajectory are implicitly made the same. If this is the case the velocity and acceleration of the first time step should be exactly zero. But we observe that the initial velocities are close to zero; not exactly zero. And the initial accelerations are not close to zero. Can you please explain the reason? 

Also, if we want to start from some given non-zero velocities and accelerations, how can we make sure that the initial velocities and accelerations of the trajectory are exactly the same as the the given values?

**Comment by balakumar-s** [comment]
cuRobo takes into account the initial velocity and acceleration by computing previous positions (before t = 0) through backward integration here: https://github.com/NVlabs/curobo/blob/3690d28c54ee4ba23e2d7cc03a89a5845e4f9546/src/curobo/curobolib/cpp/tensor_step_kernel.cu#L254

> But we observe that the initial velocities are close to zero; not exactly zero. And the initial accelerations are not close to zero

Since we are doing this backward integration, there will be numerical precision issues causing the velocity and acceleration to be not exactly zero. How far away is it from zero?

> Also, if we want to start from some given non-zero velocities and accelerations, how can we make sure that the initial velocities and accelerations of the trajectory are exactly the same as the the given values?

You can do this by setting optimize_dt to False: https://curobo.org/_api/curobo.wrap.reacher.motion_gen.html#curobo.wrap.reacher.motion_gen.MotionGenConfig.optimize_dt

This is available as an input argument to https://curobo.org/_api/curobo.wrap.reacher.motion_gen.html#curobo.wrap.reacher.motion_gen.MotionGenConfig.load_from_robot_config

**Reply by patil-apurva** [reply]
Thank you for your reply. 

The initial velocities are pretty close to zero. So initial velocities is not a concern. But the initial accelerations are far off from zero. The values of initial acceleration we get for the Kinova robot are as follows:
`2.31, 2.73, 0.99, -1.56, -0.30, 0.56, 1.11`.

As you mentioned before, the first and last three states of the trajectory are implicitly made the same. How is this achieved? Are equality constraints placed on these states, or is there an associated cost? Could you please direct me to the specific code block where these constraints or costs are set?

**Reply by balakumar-s** [reply]
The first and last three positions are implicitly made same by:
1. using start_position as the value of position for first three time steps
2. using the same index of u_position for last three time steps:
a. https://github.com/NVlabs/curobo/blob/3690d28c54ee4ba23e2d7cc03a89a5845e4f9546/src/curobo/curobolib/cpp/tensor_step_kernel.cu#L292
b. https://github.com/NVlabs/curobo/blob/3690d28c54ee4ba23e2d7cc03a89a5845e4f9546/src/curobo/curobolib/cpp/tensor_step_kernel.cu#L306

**Comment by patil-apurva** [comment]
When I set `optimize_dt`to False, I get the following error:

 `Traceback (most recent call last):
  File "/pkgs/curobo/examples/motion_gen_example_test_goal_change.py", line 180, in <module>
    demo_motion_gen()
  File "/pkgs/curobo/examples/motion_gen_example_test_goal_change.py", line 57, in demo_motion_gen
    motion_gen.warmup(parallel_finetune=True)
  File "/usr/local/lib/python3.10/dist-packages/curobo/wrap/reacher/motion_gen.py", line 1869, in warmup
    self.plan_single_js(
  File "/usr/local/lib/python3.10/dist-packages/curobo/wrap/reacher/motion_gen.py", line 2081, in plan_single_js
    result = self._plan_js_from_solve_state(
  File "/usr/local/lib/python3.10/dist-packages/curobo/wrap/reacher/motion_gen.py", line 3701, in _plan_js_from_solve_state
    opt_dt = torch.min(opt_dt[traj_result.success])
IndexError: too many indices for tensor of dimension 0`

I haven't tried to debug it myself. Is there a quick fix to remove this error?
Also, can you please explain the reason behind setting `optimize_dt` to False in order to start from non-zero velocities and accelerations?

**Reply by balakumar-s** [reply]
Is it possible to use Cartesian Pose target for planning? Joint space planning does not work with optimize_dt=False.

**Reply by balakumar-s** [reply]
We have a fix for this in the next two days. Also note that the convergence from non-static initial state will be worse as we haven't tried to improve it.
